# Data science for budiness project

dataset link
https://www.dati.lombardia.it/Attivit-Produttive/Rapporti-di-lavoro-attivati/qbau-cyuc

https://www.dati.lombardia.it/Attivit-Produttive/Rapporti-di-lavoro-cessati/nwz3-p6vm

https://www.dati.lombardia.it/Attivit-Produttive/Rapporti-di-lavoro-prorogati/chng-cman

https://www.dati.lombardia.it/Attivit-Produttive/Rapporti-di-lavoro-trasformati/8dwx-jjag

https://esploradati.censimentopopolazione.istat.it

possibili analisi dati dataset rapporti di lavoro:
1. confronti  maschi e femmine
2. divisione per fasce di età provincia e/o nazionalità in generale i valori dell colonne
3. ananlisi per anno o per mese
4. tipo di contratto in base all'età / titolo di studio / settore economico

### obbiettivo generale
rapporto tra nuovi contratti e contratti cessati per fascia di età e anno 
confrontando i vari anni per capire se i contratti sono aumentati o diminuiti (una roba del genere)

contratti per provincia: quanit contratti vengono attivati/cessati/prorogati trasformati perprovincia

### sequenza di cose da fare
1. importare dataset
2. pulire dataset e mostrare come e quali sono 
2. estrarre i valori possibili di varie colonne
3. analisi?


In [2]:
import numpy as np
import pandas as pd

In [23]:
rap_lavoro_attivati = pd.read_csv(r'dataset\Rapporti_di_lavoro_attivati.csv',parse_dates=['DATA'])
rap_lavoro_cessati = pd.read_csv(r'dataset\Rapporti_di_lavoro_cessati.csv',parse_dates=['DATA'])
rap_lavoro_prorogati = pd.read_csv(r'dataset\Rapporti_di_lavoro_prorogati.csv',parse_dates=['DATA'])
rap_lavoro_trasformati = pd.read_csv(r'dataset\Rapporti_di_lavoro_trasformati.csv',parse_dates=['DATA'])
rap_lavoro_attivati

DATA GENERE  ETA  \
0       2020-09-05      F   60   
1       2019-12-07      M   43   
2       2013-05-06      F   20   
3       2010-12-03      F   28   
4       2021-06-04      F   49   
...            ...    ...  ...   
9579789 2020-01-03      M   52   
9579790 2010-06-09      M   61   
9579791 2021-06-11      M   37   
9579792 2010-02-02      M   35   
9579793 2012-05-07      F   39   

                                 SETTOREECONOMICODETTAGLIO  \
0        Attività di famiglie e convivenze come datori ...   
1        Gestioni di funicolari, ski-lift e seggiovie s...   
2        Fabbricazione di altre apparecchiature elettri...   
3                                                 Alberghi   
4                                       Rifugi di montagna   
...                                                    ...   
9579789                       Lavori di meccanica generale   
9579790                       Lavori di meccanica generale   
9579791  Fabbricazione di parti ed accessori per bicicl...   
9579792  Fabbricazione di parti ed accessori per bicicl...   
9579793  Commercio all'ingrosso di abbigliamento e acce...   

                                              TITOLOSTUDIO  \
0                                  NESSUN TITOLO DI STUDIO   
1                                            LICENZA MEDIA   
2                                            LICENZA MEDIA   
3        DIPLOMA DI ISTRUZIONE SECONDARIA SUPERIORE  CH...   
4                                            LICENZA MEDIA   
...                                                    ...   
9579789                                      LICENZA MEDIA   
9579790                            NESSUN TITOLO DI STUDIO   
9579791                            NESSUN TITOLO DI STUDIO   
9579792                                      LICENZA MEDIA   
9579793                                      LICENZA MEDIA   

                                                 CONTRATTO MODALITALAVORO  \
0                                         LAVORO DOMESTICO    TEMPO PIENO   
1                               LAVORO A TEMPO DETERMINATO    TEMPO PIENO   
2        APPRENDISTATO PROFESSIONALIZZANTE O CONTRATTO ...    TEMPO PIENO   
3                 LAVORO INTERMITTENTE A TEMPO DETERMINATO   NON DEFINITO   
4                                     LAVORO INTERMITTENTE   NON DEFINITO   
...                                                    ...            ...   
9579789                         LAVORO A TEMPO DETERMINATO    TEMPO PIENO   
9579790                       LAVORO A TEMPO INDETERMINATO    TEMPO PIENO   
9579791                         LAVORO A TEMPO DETERMINATO    TEMPO PIENO   
9579792  LAVORO INTERINALE (O A SCOPO DI SOMMINISTRAZIO...    TEMPO PIENO   
9579793  LAVORO INTERINALE (O A SCOPO DI SOMMINISTRAZIO...            NaN   

        PROVINCIAIMPRESA ITALIANO  
0                BERGAMO  UCRAINA  
1                BERGAMO   ITALIA  
2                BERGAMO   ITALIA  
3                BERGAMO   ITALIA  
4                BERGAMO   ITALIA  
...                  ...      ...  
9579789          BERGAMO   ITALIA  
9579790          BERGAMO  ALBANIA  
9579791          BERGAMO  SENEGAL  
9579792          BERGAMO  SENEGAL  
9579793          BERGAMO   ITALIA  

[9579794 rows x 9 columns]

In [24]:
print([rap_lavoro_attivati.info(), rap_lavoro_cessati.info(),rap_lavoro_prorogati.info(),rap_lavoro_trasformati.info()])

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9579794 entries, 0 to 9579793
Data columns (total 9 columns):
 #   Column                     Dtype         
---  ------                     -----         
 0   DATA                       datetime64[ns]
 1   GENERE                     object        
 2   ETA                        int64         
 3   SETTOREECONOMICODETTAGLIO  object        
 4   TITOLOSTUDIO               object        
 5   CONTRATTO                  object        
 6   MODALITALAVORO             object        
 7   PROVINCIAIMPRESA           object        
 8   ITALIANO                   object        
dtypes: datetime64[ns](1), int64(1), object(7)
memory usage: 657.8+ MB
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3741394 entries, 0 to 3741393
Data columns (total 9 columns):
 #   Column                     Dtype 
---  ------                     ----- 
 0   DATA                       object
 1   GENERE                     object
 2   ETA                        

In [25]:
rap_lavoro_cessati.columns

Index(['DATA', 'GENERE', 'ETA', 'SETTOREECONOMICODETTAGLIO', 'TITOLOSTUDIO',
       'CONTRATTO', 'MODALITALAVORO', 'PROVINCIAIMPRESA', 'ITALIANO'],
      dtype='object')

In [26]:
rap_lavoro_prorogati.columns


Index(['DATA', 'GENERE', 'ETA', 'SETTOREECONOMICODETTAGLIO', 'TITOLOSTUDIO',
       'CONTRATTO', 'MODALITALAVORO', 'PROVINCIAIMPRESA', 'ITALIANO'],
      dtype='object')

In [27]:
rap_lavoro_trasformati.columns

Index(['DATA', 'GENERE', 'ETA', 'SETTOREECONOMICODETTAGLIO', 'TITOLOSTUDIO',
       'CONTRATTO', 'MODALITALAVORO', 'PROVINCIAIMPRESA', 'ITALIANO'],
      dtype='object')

In [53]:

def series_to_set(column, source_df):
    SET = set()
    for elem in source_df[column]:
        SET.add(elem)
    return SET

test_dataset = series_to_set('MODALITALAVORO', rap_lavoro_attivati)
test_dataset

{<NA>,
 'NON DEFINITO',
 'TEMPO PARZIALE MISTO',
 'TEMPO PARZIALE ORIZZONTALE',
 'TEMPO PARZIALE VERTICALE',
 'TEMPO PIENO'}

In [29]:
pd.DataFrame(test_dataset).info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 12 entries, 0 to 11
Data columns (total 1 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   0       12 non-null     object
dtypes: object(1)
memory usage: 224.0+ bytes


### 


altri dataset

dt1 = pd.read_csv(r'dataset\censimenti\Area_geografica_e_paesi_di_cittadinanza_Lombardia_IT1,DF_DCSS_POP.csv')
dt2 = pd.read_csv(r'dataset\censimenti\Cittadinanza_Lombardia_IT1,DF_DCSS_POP_DEMCITMIG_REG_23_ITC4,1_0.csv')
dt3 = pd.read_csv(r'dataset\censimenti\Condizione_professionale_per_cittadinanza_Lombardia_IT1,DF_DCSS.csv')
dt4 = pd.read_csv(r'dataset\censimenti\Grado_istruzione_per_cittadinanza_Lombardia_IT1,DF_DCSS_ISTR_LAV.csv')
dt6 = pd.read_csv(r'dataset\censimenti\Indicatori_demografici_Lombardia_IT1,DF_DCSS_POP_DEMCITMIG_REG_25.csv')
dt7 = pd.read_csv(r'dataset\censimenti\Numero_di_famiglie_e_popolazione_residente_in_famiglia_Lombardia.csv')
dt8 = pd.read_csv(r'dataset\censimenti\Numero_medio_di_componenti_per_famiglia_Lombardia_IT1,DF_DCSS_FAM.csv')
dt9 = pd.read_csv(r'dataset\censimenti\Popolazione_residente_per_sesso_ed_età_Lombardia_IT1,DF_DCSS_POP.csv')
dt10 = pd.read_csv(r'dataset\censimenti\Principali_paesi_di_cittadinanza_comuni_con_più_di_20_000_ab_Lombardia.csv')
dt11 = pd.read_csv(r'dataset\censimenti\Spostamenti_per_studio_o_lavoro_Lombardia_IT1,DF_DCSS_ISTR_LAV_PEN.csv')

piu' importanti:

dt5 = pd.read_csv(r'dataset\censimenti\Grado_istruzione_per_età_Lombardia_IT1,DF_DCSS_ISTR_LAV_PEN_2_REG.csv')
condPro = pd.read_csv(r'dataset\Condizione_professionale_per_età_Lombardia.csv')

questi due e' possibile fare qualche analisi in aggiuntiva considerando l'eta' come campo comune ai dataset relativi ai Raporti Di Lavoro

# Dataset analysis
In this section will be analyzed the content of the datasets. The first group of datasets contain work relations started\ended\changed\extended. Calling the *info()* method is possible to see that all dataset share the same columns that are:
- DATA: date of the contract
- GENERE: sex of teh person
- ETA: age of the person
- SETTOREECONOMICODETTAGLIO: category of work
- TITOLOSTUDIO: level of education of the person
- CONTRATTO: type of contract
- MODALITALAVORO: work mode
- PROVINCIAIMPRESA: province of the place of work?
- ITALIANO: nationality of the person

From this first look at the data is possible to see that the dataset *Rapporti_di_lavoro_attivati* is bigger than the other one and this can itroduce some sort of bias in specific type of analysis

### Changing data type
Almost all columns of the dataset are type object, in the following part we will convert every column in an appropriate type.

Starting from the column *DATA* that has been converted whilie importing the data from the csv.
Then all the other column except for *ETA* will be converted to strings

In [30]:
rap_lavoro_attivati[['GENERE','SETTOREECONOMICODETTAGLIO','TITOLOSTUDIO','CONTRATTO', 'MODALITALAVORO','PROVINCIAIMPRESA','ITALIANO']]=rap_lavoro_attivati[['GENERE','SETTOREECONOMICODETTAGLIO','TITOLOSTUDIO','CONTRATTO', 'MODALITALAVORO','PROVINCIAIMPRESA','ITALIANO']].astype('string')
rap_lavoro_cessati[['GENERE','SETTOREECONOMICODETTAGLIO','TITOLOSTUDIO','CONTRATTO', 'MODALITALAVORO','PROVINCIAIMPRESA','ITALIANO']]=rap_lavoro_cessati[['GENERE','SETTOREECONOMICODETTAGLIO','TITOLOSTUDIO','CONTRATTO', 'MODALITALAVORO','PROVINCIAIMPRESA','ITALIANO']].astype('string')
rap_lavoro_trasformati[['GENERE','SETTOREECONOMICODETTAGLIO','TITOLOSTUDIO','CONTRATTO', 'MODALITALAVORO','PROVINCIAIMPRESA','ITALIANO']]=rap_lavoro_trasformati[['GENERE','SETTOREECONOMICODETTAGLIO','TITOLOSTUDIO','CONTRATTO', 'MODALITALAVORO','PROVINCIAIMPRESA','ITALIANO']].astype('string')
rap_lavoro_prorogati[['GENERE','SETTOREECONOMICODETTAGLIO','TITOLOSTUDIO','CONTRATTO', 'MODALITALAVORO','PROVINCIAIMPRESA','ITALIANO']]=rap_lavoro_prorogati[['GENERE','SETTOREECONOMICODETTAGLIO','TITOLOSTUDIO','CONTRATTO', 'MODALITALAVORO','PROVINCIAIMPRESA','ITALIANO']].astype('string')

as is possible to see now all the column have a specific data type

In [31]:
rap_lavoro_attivati.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9579794 entries, 0 to 9579793
Data columns (total 9 columns):
 #   Column                     Dtype         
---  ------                     -----         
 0   DATA                       datetime64[ns]
 1   GENERE                     string        
 2   ETA                        int64         
 3   SETTOREECONOMICODETTAGLIO  string        
 4   TITOLOSTUDIO               string        
 5   CONTRATTO                  string        
 6   MODALITALAVORO             string        
 7   PROVINCIAIMPRESA           string        
 8   ITALIANO                   string        
dtypes: datetime64[ns](1), int64(1), string(7)
memory usage: 657.8 MB


### Searching wrong and null data
Now we will look inside the data checking null or possibly wrong value, after those data will be corrected or deleted.
First off we will find out how many null value there are in each column

In [32]:
print([rap_lavoro_attivati.isnull().sum(),rap_lavoro_cessati.isnull().sum(),rap_lavoro_prorogati.isnull().sum(),rap_lavoro_trasformati.isnull().sum()])

[DATA                              0
GENERE                            0
ETA                               0
SETTOREECONOMICODETTAGLIO      2888
TITOLOSTUDIO                    580
CONTRATTO                         0
MODALITALAVORO               440779
PROVINCIAIMPRESA                  0
ITALIANO                          0
dtype: int64, DATA                            0
GENERE                          0
ETA                             0
SETTOREECONOMICODETTAGLIO    1019
TITOLOSTUDIO                  440
CONTRATTO                       0
MODALITALAVORO                  0
PROVINCIAIMPRESA                0
ITALIANO                        0
dtype: int64, DATA                            0
GENERE                          0
ETA                             0
SETTOREECONOMICODETTAGLIO    2291
TITOLOSTUDIO                  160
CONTRATTO                       0
MODALITALAVORO                  0
PROVINCIAIMPRESA                0
ITALIANO                        0
dtype: int64, DATA                 

From this first look is possible to see that a big part of value is missing from the column *MODALITALAVORO* of the dataset *Rapporti_di_lavoro_attivati* these value needs to be replaced while the other could be simply deleted because they represent a small part of the dataset.

But before proceding now will be checked the actual value of some columnn to see if there are some non plausible data.
Starting from the column *DATA*

In [33]:
rap_lavoro_attivati['DATA'].describe(datetime_is_numeric=True)

count                          9579794
mean     2015-08-22 08:39:17.195402496
min                1973-02-01 00:00:00
25%                2012-06-07 00:00:00
50%                2015-10-08 00:00:00
75%                2018-09-07 00:00:00
max                2201-06-09 00:00:00
Name: DATA, dtype: object

In [34]:
rap_lavoro_attivati = rap_lavoro_attivati[rap_lavoro_attivati['DATA'] < np.datetime64('2022-02-20')]

In [35]:
rap_lavoro_attivati['DATA'].describe(datetime_is_numeric=True)

count                          9579793
mean     2015-08-22 08:29:05.161381888
min                1973-02-01 00:00:00
25%                2012-06-07 00:00:00
50%                2015-10-08 00:00:00
75%                2018-09-07 00:00:00
max                2021-10-06 00:00:00
Name: DATA, dtype: object

In [36]:
rap_lavoro_attivati['DATA'].max

<bound method NDFrame._add_numeric_operations.<locals>.max of 0         2020-09-05
1         2019-12-07
2         2013-05-06
3         2010-12-03
4         2021-06-04
             ...    
9579789   2020-01-03
9579790   2010-06-09
9579791   2021-06-11
9579792   2010-02-02
9579793   2012-05-07
Name: DATA, Length: 9579793, dtype: datetime64[ns]>

In [37]:
settore_economico = series_to_set('SETTOREECONOMICODETTAGLIO', rap_lavoro_attivati)
len(settore_economico)

1225

Since the `SETTOREECONOMICODETTAGLIO` column has `1125` different values belonging to it, it cannot be easily exploited to analyze data (e.g. dividing them into categories). Furthermore, there are `2888` null values in this column that can make the analysis of this feature even tougher. Considering this two facts, we can derive that dropping the whole column is more convenient than keeping it, since it introduce noise. 

In [38]:
settore_economico

{'Edizione di altri software a pacchetto (esclusi giochi per computer)',
 "Procacciatori d'affari di macchinari, impianti industriali, navi e aeromobili, macchine agricole, macchine per ufficio e computer",
 'Commercio al dettaglio di articoli per adulti (sexy shop)',
 'Commercio al dettaglio di articoli di cartoleria e forniture per ufficio',
 'Agenti e rappresentanti di bevande e prodotti similari',
 'Altre elaborazioni elettroniche di dati',
 'Fabbricazione di ascensori, montacarichi e scale mobili',
 "Commercio all'ingrosso di legname, semilavorati in legno e legno artificiale",
 'Fabbricazione di apparecchi e strumenti per odontoiatria e di apparecchi medicali per diagnosi (incluse parti staccate e accessori)',
 'Fabbricazione di alcol etilico da materiali fermentati',
 'Attività degli studi di architettura',
 'Coltivazione di legumi da granella',
 'Intermediari del commercio di motocicli e ciclomotori',
 'Fabbricazione di manici di ombrelli, bastoni e simile',
 'Agenti e rapprese

In [40]:
rap_lavoro_attivati.head()

DATA GENERE  ETA                                       TITOLOSTUDIO  \
0 2020-09-05      F   60                            NESSUN TITOLO DI STUDIO   
1 2019-12-07      M   43                                      LICENZA MEDIA   
2 2013-05-06      F   20                                      LICENZA MEDIA   
3 2010-12-03      F   28  DIPLOMA DI ISTRUZIONE SECONDARIA SUPERIORE  CH...   
4 2021-06-04      F   49                                      LICENZA MEDIA   

                                           CONTRATTO MODALITALAVORO  \
0                                   LAVORO DOMESTICO    TEMPO PIENO   
1                         LAVORO A TEMPO DETERMINATO    TEMPO PIENO   
2  APPRENDISTATO PROFESSIONALIZZANTE O CONTRATTO ...    TEMPO PIENO   
3           LAVORO INTERMITTENTE A TEMPO DETERMINATO   NON DEFINITO   
4                               LAVORO INTERMITTENTE   NON DEFINITO   

  PROVINCIAIMPRESA ITALIANO  
0          BERGAMO  UCRAINA  
1          BERGAMO   ITALIA  
2          BERGAMO   ITALIA  
3          BERGAMO   ITALIA  
4          BERGAMO   ITALIA

In the `rap_lavoro_attivati` there are a lot of null values for the column `MODALITALAVORO`, removing the rows would reduce the set of data we are analysing. Hence, we'd rather prefer to fill those null values with a suitable values. 

In [54]:
mod_lavoro = series_to_set('MODALITALAVORO', rap_lavoro_attivati)
mod_lavoro

{<NA>,
 'NON DEFINITO',
 'TEMPO PARZIALE MISTO',
 'TEMPO PARZIALE ORIZZONTALE',
 'TEMPO PARZIALE VERTICALE',
 'TEMPO PIENO'}

The value `NON DEFINITO` can be used to fill the na, since that won't introduce much bias, differently from the other available values. 

In [68]:
def clean(df):
    df.drop('SETTOREECONOMICODETTAGLIO', axis = 1, inplace = True)
    df['MODALITALAVORO'] = df['MODALITALAVORO'].fillna('NON DEFINITO')
    df.dropna(axis = 0, inplace = True)
    

clean(rap_lavoro_attivati)
rap_lavoro_attivati.isnull().sum()

C:\Users\achil\AppData\Local\Temp/ipykernel_14248/1296623547.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['MODALITALAVORO'] = df['MODALITALAVORO'].fillna('NON DEFINITO')
C:\Users\achil\anaconda3\lib\site-packages\pandas\util\_decorators.py:311: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  return func(*args, **kwargs)


DATA                0
GENERE              0
ETA                 0
TITOLOSTUDIO        0
CONTRATTO           0
MODALITALAVORO      0
PROVINCIAIMPRESA    0
ITALIANO            0
dtype: int64

A similar analysis can be made with the other three Dataframes `rap_lavoro_cessati`, `rap_lavoro_prorogati` and `rap_lavoro_trasformati`

In [70]:
clean(rap_lavoro_cessati)
clean(rap_lavoro_prorogati)
clean(rap_lavoro_trasformati)